File: keyorix.js
'use strict';

/**
 * Keyorix Node.js SDK
 * Zero external dependencies — uses Node.js built-in https/http modules.
 *
 * Quick start:
 *
 *   const keyorix = require('keyorix');
 *
 *   const token = await keyorix.login('http://your-server:8080', 'admin', 'password');
 *   const client = new keyorix.Client('http://your-server:8080', token);
 *
 *   const dbPassword = await client.getSecret('db-password', 'production');
 *   const secrets = await client.listSecrets('production');
 */

const http = require('http');
const https = require('https');

// ── Errors ───────────────────────────────────────────────────────────────────

class KeyorixError extends Error {
  constructor(message) {
    super(message);
    this.name = 'KeyorixError';
  }
}

class AuthError extends KeyorixError {
  constructor(message) {
    super(message);
    this.name = 'AuthError';
  }
}

class SecretNotFoundError extends KeyorixError {
  constructor(message) {
    super(message);
    this.name = 'SecretNotFoundError';
  }
}

// ── HTTP helper ──────────────────────────────────────────────────────────────

function request(options, body) {
  return new Promise((resolve, reject) => {
    const transport = options.protocol === 'https:' ? https : http;
    const req = transport.request(options, (res) => {
      const chunks = [];
      res.on('data', (chunk) => chunks.push(chunk));
      res.on('end', () => {
        const raw = Buffer.concat(chunks).toString();
        resolve({ status: res.statusCode, body: raw });
      });
    });
    req.on('error', reject);
    if (body) req.write(body);
    req.end();
  });
}

function parseUrl(serverUrl) {
  const u = new URL(serverUrl);
  return {
    protocol: u.protocol,
    hostname: u.hostname,
    port: u.port || (u.protocol === 'https:' ? 443 : 80),
  };
}

// ── login ────────────────────────────────────────────────────────────────────

/**
 * Authenticate with Keyorix and return a session token.
 *
 * @param {string} serverUrl - Base URL of your Keyorix server
 * @param {string} username
 * @param {string} password
 * @param {number} [timeout=30000] - Timeout in milliseconds
 * @returns {Promise<string>} Session token
 */
async function login(serverUrl, username, password, timeout = 30000) {
  const base = parseUrl(serverUrl);
  const body = JSON.stringify({ username, password });

  const options = {
    ...base,
    path: '/auth/login',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(body),
    },
    timeout,
  };

  let resp;
  try {
    resp = await request(options, body);
  } catch (err) {
    throw new KeyorixError(`Server unreachable: ${err.message}`);
  }

  if (resp.status !== 200) {
    throw new AuthError(`Login failed (HTTP ${resp.status}): ${resp.body}`);
  }

  const data = JSON.parse(resp.body);
  const token = data?.data?.token;
  if (!token) throw new AuthError('No token in login response');
  return token;
}

// ── Client ───────────────────────────────────────────────────────────────────

class Client {
  /**
   * @param {string} serverUrl - Base URL of your Keyorix server
   * @param {string} token - Session token
   * @param {object} [opts]
   * @param {number} [opts.timeout=30000] - Timeout in milliseconds
   */
  constructor(serverUrl, token, opts = {}) {
    this._base = serverUrl.replace(/\/$/, '');
    this._token = token;
    this._timeout = opts.timeout || 30000;
    this._parsed = parseUrl(this._base);
  }

  async _request(path) {
    const options = {
      ...this._parsed,
      path,
      method: 'GET',
      headers: { Authorization: `Bearer ${this._token}` },
      timeout: this._timeout,
    };

    let resp;
    try {
      resp = await request(options);
    } catch (err) {
      throw new KeyorixError(`Request failed: ${err.message}`);
    }

    if (resp.status === 401) throw new AuthError('Unauthorized — check your token');
    if (resp.status !== 200) {
      throw new KeyorixError(`Server returned ${resp.status}: ${resp.body}`);
    }

    return JSON.parse(resp.body);
  }

  /**
   * Check if the server is reachable and healthy.
   * @returns {Promise<boolean>}
   */
  async health() {
    const options = {
      ...this._parsed,
      path: '/health',
      method: 'GET',
      timeout: this._timeout,
    };
    try {
      const resp = await request(options);
      return resp.status === 200;
    } catch (err) {
      throw new KeyorixError(`Server unreachable: ${err.message}`);
    }
  }

  /**
   * List secrets visible to the authenticated user.
   * @param {string} [environment] - Filter by environment name
   * @returns {Promise<Array>}
   */
  async listSecrets(environment = '') {
    let path = '/api/v1/secrets';
    if (environment) path += `?environment=${encodeURIComponent(environment)}`;
    const data = await this._request(path);
    return (data?.data?.secrets || []).map((s) => ({
      id: s.ID,
      name: s.Name,
      type: s.Type,
      environment: s.environment_name,
      namespace: s.namespace_name,
      createdAt: s.CreatedAt,
    }));
  }

  /**
   * Get the value of a secret by name.
   * @param {string} name - Secret name
   * @param {string} [environment] - Environment to search in
   * @returns {Promise<string>} Plaintext secret value
   */
  async getSecret(name, environment = '') {
    const secrets = await this.listSecrets(environment);
    const secret = secrets.find((s) => s.name === name);
    if (!secret) {
      const envMsg = environment ? ` in environment '${environment}'` : '';
      throw new SecretNotFoundError(`Secret '${name}' not found${envMsg}`);
    }
    return this._getSecretValue(secret.id);
  }

  async _getSecretValue(secretId) {
    const data = await this._request(`/api/v1/secrets/${secretId}?include_value=true`);
    return data?.data?.value || '';
  }
}

module.exports = { Client, login, KeyorixError, AuthError, SecretNotFoundError };

File: keyorix.py
"""
keyorix — Python client for the Keyorix secrets manager.

Quick start:

    import keyorix

    # Option 1: use a token directly
    client = keyorix.Client("http://your-server:8080", "your-token")

    # Option 2: log in with username/password
    token = keyorix.login("http://your-server:8080", "admin", "password")
    client = keyorix.Client("http://your-server:8080", token)

    # Get a secret
    db_password = client.get_secret("db-password", "production")

    # List secrets
    secrets = client.list_secrets("production")
"""

import json
import urllib.request
import urllib.error
from dataclasses import dataclass
from datetime import datetime
from typing import List, Optional


class KeyorixError(Exception):
    """Raised when the Keyorix API returns an error."""

    pass


class AuthError(KeyorixError):
    """Raised on authentication failure."""

    pass


class SecretNotFoundError(KeyorixError):
    """Raised when a secret cannot be found."""

    pass


@dataclass
class Secret:
    """A secret returned by the Keyorix API."""

    id: int
    name: str
    type: str
    environment: str
    namespace: str
    created_at: str

    @classmethod
    def _from_dict(cls, data: dict) -> "Secret":
        return cls(
            id=data.get("ID", 0),
            name=data.get("Name", ""),
            type=data.get("Type", ""),
            environment=data.get("environment_name", ""),
            namespace=data.get("namespace_name", ""),
            created_at=data.get("CreatedAt", ""),
        )


def login(server_url: str, username: str, password: str, timeout: int = 30) -> str:
    """Authenticate with Keyorix and return a session token.

    Args:
        server_url: Base URL of your Keyorix server (e.g. "http://localhost:8080")
        username: Keyorix username
        password: Keyorix password
        timeout: Request timeout in seconds (default 30)

    Returns:
        Session token string

    Raises:
        AuthError: If authentication fails
        KeyorixError: On other errors
    """
    payload = json.dumps({"username": username, "password": password}).encode()
    req = urllib.request.Request(
        f"{server_url.rstrip('/')}/auth/login",
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            data = json.loads(resp.read())
            token = data.get("data", {}).get("token", "")
            if not token:
                raise AuthError("No token in login response")
            return token
    except urllib.error.HTTPError as e:
        body = e.read().decode(errors="replace")
        raise AuthError(f"Login failed (HTTP {e.code}): {body}") from e
    except urllib.error.URLError as e:
        raise KeyorixError(f"Server unreachable: {e.reason}") from e


class Client:
    """Keyorix API client.

    Args:
        server_url: Base URL of your Keyorix server
        token: Session token (obtain via keyorix.login() or CLI)
        timeout: Request timeout in seconds (default 30)
    """

    def __init__(self, server_url: str, token: str, timeout: int = 30):
        self._base = server_url.rstrip("/")
        self._token = token
        self._timeout = timeout

    def _request(self, method: str, path: str) -> dict:
        req = urllib.request.Request(
            f"{self._base}{path}",
            headers={"Authorization": f"Bearer {self._token}"},
            method=method,
        )
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            if e.code == 401:
                raise AuthError("Unauthorized — check your token") from e
            body = e.read().decode(errors="replace")
            raise KeyorixError(f"Server returned {e.code}: {body}") from e
        except urllib.error.URLError as e:
            raise KeyorixError(f"Request failed: {e.reason}") from e

    def health(self) -> bool:
        """Check if the server is reachable and healthy.

        Returns:
            True if healthy

        Raises:
            KeyorixError: If server is unreachable or unhealthy
        """
        req = urllib.request.Request(
            f"{self._base}/health",
            method="GET",
        )
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                return resp.status == 200
        except (urllib.error.HTTPError, urllib.error.URLError) as e:
            raise KeyorixError(f"Server unreachable: {e}") from e

    def list_secrets(self, environment: str = "") -> List[Secret]:
        """List secrets visible to the authenticated user.

        Args:
            environment: Filter by environment ("production", "staging", "development").
                         Pass empty string for all environments.

        Returns:
            List of Secret objects
        """
        path = "/api/v1/secrets"
        if environment:
            path += f"?environment={urllib.parse.quote(environment)}"
        data = self._request("GET", path)
        secrets_data = data.get("data", {}).get("secrets", [])
        return [Secret._from_dict(s) for s in secrets_data]

    def get_secret(self, name: str, environment: str = "") -> str:
        """Get the value of a secret by name.

        Args:
            name: Secret name
            environment: Environment to search in ("production", "staging", "development")

        Returns:
            Plaintext secret value

        Raises:
            SecretNotFoundError: If secret is not found
            KeyorixError: On other errors
        """
        secrets = self.list_secrets(environment)
        for secret in secrets:
            if secret.name == name:
                return self._get_secret_value(secret.id)
        env_msg = f" in environment {environment!r}" if environment else ""
        raise SecretNotFoundError(f"Secret {name!r} not found{env_msg}")

    def _get_secret_value(self, secret_id: int) -> str:
        data = self._request("GET", f"/api/v1/secrets/{secret_id}?include_value=true")
        return data.get("data", {}).get("value", "")


# Fix missing import
import urllib.parse
